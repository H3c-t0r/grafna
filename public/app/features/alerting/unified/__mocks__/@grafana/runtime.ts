const original = jest.requireActual<typeof import('@grafana/runtime')>('@grafana/runtime');

module.exports = {
  __esModule: true,
  ...original,
  useReturnToPrevious: jest.fn(),
  getPluginLinkExtensions: jest.fn(),
  usePluginLinkExtensions: jest.fn(),
};
