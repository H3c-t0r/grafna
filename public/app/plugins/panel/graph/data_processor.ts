import _ from 'lodash';
import {
  LegacyResponseData,
  TimeRange,
  colors,
  getColorFromHexRgbOrName,
  FieldCache,
  FieldType,
  SeriesData,
  Field,
} from '@grafana/ui';
import TimeSeries from 'app/core/time_series2';
import config from 'app/core/config';
import { getProcessedSeriesData } from 'app/features/dashboard/state/PanelQueryState';

type Options = {
  dataList: LegacyResponseData[];
  range?: TimeRange;
};

export class DataProcessor {
  constructor(private panel) {}

  getSeriesList(options: Options): TimeSeries[] {
    if (!options.dataList || options.dataList.length === 0) {
      return [];
    }

    switch (this.panel.xaxis.mode) {
      case 'histogram': {
        let histogramDataList: SeriesData[];
        if (this.panel.stack) {
          histogramDataList = options.dataList;
        } else {
          histogramDataList = [
            {
              name: 'count',
              fields: [{ name: 'Value' }, { name: 'time', type: FieldType.time }],
              rows: _.concat([], _.flatten(_.map(options.dataList, 'rows'))),
            },
          ];
        }

        return this.getTimeSeries(histogramDataList, options);
      }
      case 'series':
      default: {
        return this.getTimeSeries(options.dataList, options);
      }
    }

    return [];
  }

  setPanelDefaultsForNewXAxisMode() {
    switch (this.panel.xaxis.mode) {
      case 'time': {
        this.panel.bars = false;
        this.panel.lines = true;
        this.panel.points = false;
        this.panel.legend.show = true;
        this.panel.tooltip.shared = true;
        this.panel.xaxis.values = [];
        break;
      }
      case 'series': {
        this.panel.bars = true;
        this.panel.lines = false;
        this.panel.points = false;
        this.panel.stack = false;
        this.panel.legend.show = false;
        this.panel.tooltip.shared = false;
        this.panel.xaxis.values = ['total'];
        break;
      }
      case 'histogram': {
        this.panel.bars = true;
        this.panel.lines = false;
        this.panel.points = false;
        this.panel.stack = false;
        this.panel.legend.show = false;
        this.panel.tooltip.shared = false;
        break;
      }
    }
  }

  getTimeSeries(seriesData: LegacyResponseData[], options: Options) {
    const list: TimeSeries[] = [];
    for (const series of getProcessedSeriesData(seriesData)) {
      const { fields } = series;
      const cache = new FieldCache(fields);
      const time = cache.getFirstFieldOfType(FieldType.time);
      if (!time) {
        continue;
      }

      const seriesName = series.name ? series.name : series.refId;

      for (let i = 0; i < fields.length; i++) {
        if (fields[i].type !== FieldType.number) {
          continue;
        }
        const field = fields[i];
        let name = field.title;
        if (!field.title) {
          name = field.name;
        }
        if (seriesName && seriesData.length > 0 && name !== seriesName) {
          name = seriesName + ' ' + name;
        }

        list.push(
          this.toTimeSeries(
            field,
            name, // Alias
            series.rows.map(row => {
              return [row[i], row[time.index]];
            }),
            list.length,
            options.range
          )
        );
      }
    }
    return list;
  }

  private toTimeSeries(field: Field, alias: string, datapoints: any[][], index: number, range?: TimeRange) {
    const colorIndex = index % colors.length;
    const color = this.panel.aliasColors[alias] || colors[colorIndex];

    const series = new TimeSeries({
      datapoints: datapoints || [],
      alias: alias,
      color: getColorFromHexRgbOrName(color, config.theme.type),
      unit: field.unit,
    });

    if (datapoints && datapoints.length > 0 && range) {
      const last = datapoints[datapoints.length - 1][1];
      const from = range.from;

      if (last - from.valueOf() < -10000) {
        series.isOutsideRange = true;
      }
    }
    return series;
  }

  validateXAxisSeriesValue() {
    switch (this.panel.xaxis.mode) {
      case 'series': {
        if (this.panel.xaxis.values.length === 0) {
          this.panel.xaxis.values = ['total'];
          return;
        }

        const validOptions = this.getXAxisValueOptions({});
        const found: any = _.find(validOptions, { value: this.panel.xaxis.values[0] });
        if (!found) {
          this.panel.xaxis.values = ['total'];
        }
        return;
      }
    }
  }

  getXAxisValueOptions(options) {
    switch (this.panel.xaxis.mode) {
      case 'series': {
        return [
          { text: 'Avg', value: 'avg' },
          { text: 'Min', value: 'min' },
          { text: 'Max', value: 'max' },
          { text: 'Total', value: 'total' },
          { text: 'Count', value: 'count' },
        ];
      }
    }

    return [];
  }

  pluckDeep(obj: any, property: string) {
    const propertyParts = property.split('.');
    let value = obj;
    for (let i = 0; i < propertyParts.length; ++i) {
      if (value[propertyParts[i]]) {
        value = value[propertyParts[i]];
      } else {
        return undefined;
      }
    }
    return value;
  }
}
