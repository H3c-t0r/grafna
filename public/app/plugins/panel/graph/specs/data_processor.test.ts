import { DataProcessor } from '../data_processor';
import { LegacyResponseData } from '@grafana/ui';

describe('Graph DataProcessor', () => {
  const panel: any = {
    xaxis: {},
    aliasColors: {},
  };

  const processor = new DataProcessor(panel);

  describe('Given default xaxis options and query that returns docs', () => {
    beforeEach(() => {
      panel.xaxis.mode = 'time';
      panel.xaxis.name = 'hostname';
      panel.xaxis.values = [];

      processor.getSeriesList({
        dataList: [
          {
            type: 'docs',
            datapoints: [{ hostname: 'server1', avg: 10 }],
          },
        ],
      });
    });

    it('Should automatically set xaxis mode to field', () => {
      expect(panel.xaxis.mode).toBe('field');
    });
  });

  describe('getDataFieldNames(', () => {
    const dataList = [
      {
        type: 'docs',
        datapoints: [
          {
            hostname: 'server1',
            valueField: 11,
            nested: {
              prop1: 'server2',
              value2: 23,
            },
          },
        ],
      },
    ];

    it('Should return all field names', () => {
      const fields = processor.getDataFieldNames(dataList, false);
      expect(fields).toContain('hostname');
      expect(fields).toContain('valueField');
      expect(fields).toContain('nested.prop1');
      expect(fields).toContain('nested.value2');
    });

    it('Should return all number fields', () => {
      const fields = processor.getDataFieldNames(dataList, true);
      expect(fields).toContain('valueField');
      expect(fields).toContain('nested.value2');
    });
  });

  describe('getTimeSeries from LegacyResponseData', () => {
    // Try each type of data
    const dataList = [
      {
        alias: 'First (time_series)',
        datapoints: [[1, 1001], [2, 1002], [3, 1003]],
        unit: 'watt',
      },
      {
        name: 'table_data',
        columns: [
          { text: 'time' },
          { text: 'v1', unit: 'ohm' },
          { text: 'v2' }, // no unit
          { text: 'string' }, // skipped
        ],
        rows: [
          [1001, 0.1, 1.1, 'a'], // a
          [1002, 0.2, 2.2, 'b'], // b
          [1003, 0.3, 3.3, 'c'], // c
        ],
      },
      {
        name: 'series',
        fields: [
          { name: 'v1' }, // first
          { name: 'v2' }, // second
          { name: 'string' }, // skip
          { name: 'time' }, // Time is last column
        ],
        rows: [[0.1, 1.1, 'a', 1001], [0.2, 2.2, 'b', 1002], [0.3, 3.3, 'c', 1003]],
      },
    ] as LegacyResponseData[];

    it('Should return a new series for each field', () => {
      const series = processor.getTimeSeries(dataList, { dataList: [] });
      expect(series.length).toEqual(5);
      expect(series).toMatchSnapshot();
    });
  });
});
