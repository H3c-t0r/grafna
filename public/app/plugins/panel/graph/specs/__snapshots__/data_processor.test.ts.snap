// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Graph DataProcessor getTimeSeries from LegacyResponseData Should return a new series for each field 1`] = `
Array [
  TimeSeries {
    "alias": "Value",
    "aliasEscaped": "Value",
    "bars": Object {
      "fillColor": "#7EB26D",
    },
    "color": "#7EB26D",
    "datapoints": Array [
      Array [
        1,
        1001,
      ],
      Array [
        2,
        1002,
      ],
      Array [
        3,
        1003,
      ],
    ],
    "fieldName": "Value",
    "hasMsResolution": false,
    "id": "Value",
    "label": "Value",
    "legend": true,
    "stats": Object {},
    "unit": "watt",
    "valueFormater": [Function],
    "view": Array [
      Object {
        "Time": 1001,
        "Value": 1,
      },
      Object {
        "Time": 1002,
        "Value": 2,
      },
      Object {
        "Time": 1003,
        "Value": 3,
      },
    ],
  },
  TimeSeries {
    "alias": "table_data v1",
    "aliasEscaped": "table_data v1",
    "bars": Object {
      "fillColor": "#EAB839",
    },
    "color": "#EAB839",
    "datapoints": Array [
      Array [
        0.1,
        1001,
      ],
      Array [
        0.2,
        1002,
      ],
      Array [
        0.3,
        1003,
      ],
    ],
    "fieldName": "v1",
    "hasMsResolution": false,
    "id": "table_data v1",
    "label": "table_data v1",
    "legend": true,
    "stats": Object {},
    "unit": "ohm",
    "valueFormater": [Function],
    "view": Array [
      Object {
        "string": "a",
        "time": 1001,
        "v1": 0.1,
        "v2": 1.1,
      },
      Object {
        "string": "b",
        "time": 1002,
        "v1": 0.2,
        "v2": 2.2,
      },
      Object {
        "string": "c",
        "time": 1003,
        "v1": 0.3,
        "v2": 3.3,
      },
    ],
  },
  TimeSeries {
    "alias": "table_data v2",
    "aliasEscaped": "table_data v2",
    "bars": Object {
      "fillColor": "#6ED0E0",
    },
    "color": "#6ED0E0",
    "datapoints": Array [
      Array [
        1.1,
        1001,
      ],
      Array [
        2.2,
        1002,
      ],
      Array [
        3.3,
        1003,
      ],
    ],
    "fieldName": "v2",
    "hasMsResolution": false,
    "id": "table_data v2",
    "label": "table_data v2",
    "legend": true,
    "stats": Object {},
    "unit": undefined,
    "valueFormater": [Function],
    "view": Array [
      Object {
        "string": "a",
        "time": 1001,
        "v1": 0.1,
        "v2": 1.1,
      },
      Object {
        "string": "b",
        "time": 1002,
        "v1": 0.2,
        "v2": 2.2,
      },
      Object {
        "string": "c",
        "time": 1003,
        "v1": 0.3,
        "v2": 3.3,
      },
    ],
  },
  TimeSeries {
    "alias": "series v1",
    "aliasEscaped": "series v1",
    "bars": Object {
      "fillColor": "#EF843C",
    },
    "color": "#EF843C",
    "datapoints": Array [
      Array [
        0.1,
        1001,
      ],
      Array [
        0.2,
        1002,
      ],
      Array [
        0.3,
        1003,
      ],
    ],
    "fieldName": "v1",
    "hasMsResolution": false,
    "id": "series v1",
    "label": "series v1",
    "legend": true,
    "stats": Object {},
    "unit": undefined,
    "valueFormater": [Function],
    "view": Array [
      Object {
        "string": "a",
        "time": 1001,
        "v1": 0.1,
        "v2": 1.1,
      },
      Object {
        "string": "b",
        "time": 1002,
        "v1": 0.2,
        "v2": 2.2,
      },
      Object {
        "string": "c",
        "time": 1003,
        "v1": 0.3,
        "v2": 3.3,
      },
    ],
  },
  TimeSeries {
    "alias": "series v2",
    "aliasEscaped": "series v2",
    "bars": Object {
      "fillColor": "#E24D42",
    },
    "color": "#E24D42",
    "datapoints": Array [
      Array [
        1.1,
        1001,
      ],
      Array [
        2.2,
        1002,
      ],
      Array [
        3.3,
        1003,
      ],
    ],
    "fieldName": "v2",
    "hasMsResolution": false,
    "id": "series v2",
    "label": "series v2",
    "legend": true,
    "stats": Object {},
    "unit": undefined,
    "valueFormater": [Function],
    "view": Array [
      Object {
        "string": "a",
        "time": 1001,
        "v1": 0.1,
        "v2": 1.1,
      },
      Object {
        "string": "b",
        "time": 1002,
        "v1": 0.2,
        "v2": 2.2,
      },
      Object {
        "string": "c",
        "time": 1003,
        "v1": 0.3,
        "v2": 3.3,
      },
    ],
  },
]
`;

exports[`Graph DataProcessor getTimeSeries from LegacyResponseData Should return single histogram 1`] = `
Array [
  TimeSeries {
    "alias": "Count",
    "aliasEscaped": "Count",
    "bars": Object {
      "fillColor": "#7EB26D",
    },
    "color": "#7EB26D",
    "datapoints": Array [
      Array [
        1,
        1001,
      ],
      Array [
        2,
        1002,
      ],
      Array [
        3,
        1003,
      ],
      Array [
        0.1,
        1001,
      ],
      Array [
        0.2,
        1002,
      ],
      Array [
        0.3,
        1003,
      ],
      Array [
        1.1,
        1001,
      ],
      Array [
        2.2,
        1002,
      ],
      Array [
        3.3,
        1003,
      ],
      Array [
        0.1,
        1001,
      ],
      Array [
        0.2,
        1002,
      ],
      Array [
        0.3,
        1003,
      ],
      Array [
        1.1,
        1001,
      ],
      Array [
        2.2,
        1002,
      ],
      Array [
        3.3,
        1003,
      ],
    ],
    "fieldName": "Value",
    "hasMsResolution": false,
    "id": "Value",
    "label": "Value",
    "legend": true,
    "stats": Object {},
    "unit": "watt",
    "valueFormater": [Function],
    "view": Array [
      Object {
        "Time": 1001,
        "Value": 1,
      },
      Object {
        "Time": 1002,
        "Value": 2,
      },
      Object {
        "Time": 1003,
        "Value": 3,
      },
    ],
  },
]
`;
